### DISCUSSION TITLE
Custom elements: documentation, examples, tests

### DISCUSSION THREAD

**Discussion by tony** [facebook/docusaurus#5614] [Ideas]
I have a docusaurus `2.0.0 beta-0` project ([website](https://social-embed.git-pull.com), [source](https://github.com/social-embed/social-embed/tree/%40social-embed/wc%400.0.3-next.5/packages/site)) that works with a [custom element](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements)

Unfortunately, moving from [beta 0 to beta 1](https://github.com/facebook/docusaurus/pull/4789#issuecomment-886044485) with code that worked before. 

Related: https://github.com/facebook/docusaurus/issues/5416

If would be great if docusaurus had a canonical way to add custom elements and tests to verify integrating Custom elements work and don't break

P.S. Thank you so much for this project!  Beautiful, fast documentation. This is the best doc engine on the planet at the moment.

#### minimal example

https://github.com/tony/docusaurus-custom-element-issue

- :white_check_mark: docusaurus 2.0.0-beta.0:  [`2.0.0-beta.0`](https://github.com/tony/docusaurus-custom-element-issue/tree/2.0.0-beta.0)
  https://2-0-0-beta-0.dlpncn4sz6ns6.amplifyapp.com/
- 🚫 docusaurus 2.0.0-beta.1:  [`2.0.0-beta.1`](https://github.com/tony/docusaurus-custom-element-issue/tree/2.0.0-beta.1)
  https://2-0-0-beta-1.dlpncn4sz6ns6.amplifyapp.com/
- 🚫 docusaurus 2.0.0-beta.6:  [`2.0.0-beta.6`](https://github.com/tony/docusaurus-custom-element-issue/tree/2.0.0-beta.6)
  https://2-0-0-beta-6.dlpncn4sz6ns6.amplifyapp.com/
- 🚫 docusaurus canary:  [`canary`](https://github.com/tony/docusaurus-custom-element-issue/tree/canary)
  https://canary.dlpncn4sz6ns6.amplifyapp.com/

#### social-embed (monorepo)

beta 1: https://github.com/social-embed/social-embed/pull/36 ([diff](https://github.com/social-embed/social-embed/pull/36/files))
canary: https://github.com/social-embed/social-embed/pull/37 ([diff](https://github.com/social-embed/social-embed/pull/37/files))

- 🚫 Live code blocks with custom element

   beta 0 (works): https://social-embed.git-pull.com/
   beta 1 (breaks): https://docusaurus-2-0-0-beta-1.d1bxvdq53pbygo.amplifyapp.com/
   canary (breaks): https://docusaurus-2-0-0-next.d1bxvdq53pbygo.amplifyapp.com/
- ✔️ Normal custom elements in JSX

  beta 0 (works): https://social-embed.git-pull.com/docs/wc/providers/vimeo/
  beta 1 (works): https://docusaurus-2-0-0-beta-1.d1bxvdq53pbygo.amplifyapp.com/docs/wc/providers/vimeo/
  `docusaurus 2.0.0@canary`  (works):
   canary (works): ttps://docusaurus-2-0-0-next.d1bxvdq53pbygo.amplifyapp.com/docs/wc/providers/vimeo/

**Comment by slorber** [comment]
Thanks

I don't know much about custom elements, please explain how you made the integration on your site.

Also worth testing on the canary release

**Reply by tony** [reply]
Thank you for checking this our @slorber! I am very grateful!

I will check out the canary release this weekend!

Here's how it's working on beta 0:

- [web component project](https://github.com/social-embed/social-embed/tree/%40social-embed/wc%400.0.4-next.0/packages/wc), [OEmbedElement.ts](https://github.com/social-embed/social-embed/blob/%40social-embed/wc%400.0.4-next.0/packages/wc/src/OEmbedElement.ts)

  A [LitElement](https://github.com/lit/lit) project to create a HTML standard [Custom Element](https://html.spec.whatwg.org/multipage/custom-elements.html) ([MDN](https://developer.mozilla.org/en-US/docs/Web/Web_Components/Using_custom_elements))

  1. Built via `tsc` ([package.json](https://github.com/social-embed/social-embed/blob/%40social-embed/wc%400.0.4-next.0/packages/wc/package.json#L11))
      For normal sites, publishing this as a package is adequate, but for a static site, with no access to `window`, we need a UMD that temporarily handles the absence of `window` in buildtime.
  2. For docusaurus compatibility, the component's `tsc` output is additionally compiled into a single UMD file via `rollup` (below)
- [docusaurus project](https://github.com/social-embed/social-embed/tree/master/packages/site)
  - [package.json](https://github.com/social-embed/social-embed/blob/%40social-embed/wc%400.0.4-next.0/packages/site/package.json)'s [`docs:oembed`](https://github.com/social-embed/social-embed/blob/%40social-embed/wc%400.0.4-next.0/packages/site/package.json#L21) scripts rolls it up into a UMD ([rollup config](https://github.com/social-embed/social-embed/blob/%40social-embed/wc%400.0.4-next.0/packages/site/rollup.config.js))
  - [docusaurus.config.js](https://github.com/social-embed/social-embed/blob/%40social-embed/wc%400.0.4-next.0/packages/site/docusaurus.config.js)'s loads `o-embed.bundled.js` [as a script](https://github.com/social-embed/social-embed/blob/%40social-embed/wc%400.0.4-next.0/packages/site/docusaurus.config.js#L14)
    ![image](https://user-images.githubusercontent.com/26336/135109821-e55bb3e0-c12e-4635-9544-e696f23a5f85.png)
    Above: https://social-embed.git-pull.com right now

When I'm getting the error in beta 1+, it seems somehow babel, the only thing I think could be different is babel / another JS tool trying to handle `scripts`. 

Will follow up soon (weekend most likely)

**Reply by tony** [reply]
Follow up. It's live code blocks.

I added it to the [discussion topic](https://github.com/facebook/docusaurus/discussions/5614#discussion-3592184).

**Reply by tony** [reply]
Actually, just did an mvp right now:

- :white_check_mark: docusaurus 2.0.0-beta.0:  [`2.0.0-beta.0`](https://github.com/tony/docusaurus-custom-element-issue/tree/2.0.0-beta.0)
  https://2-0-0-beta-0.dlpncn4sz6ns6.amplifyapp.com/
- 🚫 docusaurus 2.0.0-beta.1:  [`2.0.0-beta.1`](https://github.com/tony/docusaurus-custom-element-issue/tree/2.0.0-beta.1)
  https://2-0-0-beta-1.dlpncn4sz6ns6.amplifyapp.com/
- 🚫 docusaurus 2.0.0-beta.6:  [`2.0.0-beta.6`](https://github.com/tony/docusaurus-custom-element-issue/tree/2.0.0-beta.6)
  https://2-0-0-beta-6.dlpncn4sz6ns6.amplifyapp.com/
- 🚫 docusaurus canary  [`canary`](https://github.com/tony/docusaurus-custom-element-issue/tree/canary)
  https://canary.dlpncn4sz6ns6.amplifyapp.com/

**Reply by tony** [reply]
@slorber is there any more information I can provide on this (custom elements in live code blocks)?   The above format has the source code and static websites for each branch to demonstrate what happens

**Reply by slorber** [reply]
@tony unfortunately I can't make this a priority atm, and I'm not skilled in web components.

This issue is hard to scan, it contains so many links and I don't even really know what to look for on each link. 

I do see the player failing in a react-live playground though.

Have you tried a minimalist custom element (not `<o-embed>`?). 

Have you tried using it outside of the live playground?

I suggest trying to wrap your custom element rendering inside a `<BrowserOnly>` component, it's possibly related to React hydration.

I'd suggest submitting a PR to our own website showing how to integrate a very basic custom element on a page (doesn't matter where for now, we can figure this out). This way we can more easily inspect the diff and eventually work on it.

**Reply by tony** [reply]
I am very grateful for the response. My dream is to use the latest and greatest docusaurus on this site. 

> This issue is hard to scan, it contains so many links and I don't even really know what to look for on each link.
> I do see the player failing in a react-live playground though.

It's only the first page's react-live playground:
![image](https://user-images.githubusercontent.com/26336/146958017-9b9dbd81-ed88-420d-8014-b8f427834199.png)

The rest of the links demonstrate `<o-embed>` works correctly in the mdx (excellent!)

> Have you tried a minimalist custom element (not `<o-embed>`?).

I will.

> Have you tried using it outside of the live playground?

It works ([1](https://canary.dlpncn4sz6ns6.amplifyapp.com/), [2](https://canary.dlpncn4sz6ns6.amplifyapp.com/docs/wc/providers/vimeo/)):
![image](https://user-images.githubusercontent.com/26336/146958263-e2018631-54b6-4143-8386-e82f808f69a5.png)

> I suggest trying to wrap your custom element rendering inside a <BrowserOnly> component, it's possibly related to React hydration.

Just `react-live` as of now.  I assume `<BrowserOnly>` isn't an issue there?

> I'd suggest submitting a PR to our own website showing how to integrate a very basic custom element on a page (doesn't matter where for now, we can figure this out). This way we can more easily inspect the diff and eventually work on it.

I will follow through with this and it will greatly enhance the experience on this front.

**Reply by slorber** [reply]
I believe this issue has been fixed already in https://github.com/facebook/docusaurus/pull/5754

Please try to upgrade to the latest version and see if it works better

**Reply by tony** [reply]
@slorber Thank you. I will update the canary release and add a beta 13 one ASAP

**Reply by slorber** [reply]
👍 

beta 14 is being released in 5 min

**Reply by tony** [reply]
canary, beta 13, and beta 14. It will be my lunch :)

**Reply by tony** [reply]
@slorber Early preview of docusaurus beta 14 (compression enabled)

https://docusaurus-2-0-0-beta-14.d1bxvdq53pbygo.amplifyapp.com/ [src](https://github.com/social-embed/social-embed/tree/docusaurus-2.0.0-beta.14/packages/site)

![image](https://user-images.githubusercontent.com/26336/146965646-ec0d63cb-9db4-4604-ade7-990678f699dc.png)

Live editor now loads but is frozen.


<details>
<summary>Side note, this other console issue may be unrelated since this shows even on pages without live editor</summary>

![image](https://user-images.githubusercontent.com/26336/146965715-760eb7c4-4185-4702-a36b-f610176cf8a5.png)


```
Uncaught (in promise) ReferenceError: exports is not defined
```
</details>

I will spin up an uncompressed version during lunch

**Reply by tony** [reply]
Unminified version: https://docusaurus-2-0-0-beta-14-no-minify.d1bxvdq53pbygo.amplifyapp.com/, [src](https://github.com/social-embed/social-embed/tree/docusaurus-2.0.0-beta.14-no-minify/packages/site)

One suspect thing is the [`babel.config.js`](https://github.com/social-embed/social-embed/blob/docusaurus-2.0.0-beta.14-no-minify/packages/site/babel.config.js) using [@babel/plugin-transform-modules-commonjs](https://babeljs.io/docs/en/babel-plugin-transform-modules-commonjs) since going from [alpha.72 to beta.0](https://github.com/social-embed/social-embed/commit/ce6e0667a66c9d55091f8bf34854ad70f9ba91e1). I will look into that.

**Reply by slorber** [reply]
It looks to me that the problem is not related in any way to custom elements, it's just that React fails to hydrate because the runtime JS bundle fails to load.

You can see that you can't even collapse the sidebar categories or use the search or toggle the dark mode

I don't know why this happens, maybe it's due to `copy-text-to-clipboard` being hoisted due to being used inside a monorepo 🤷‍♂️

**Reply by tony** [reply]
@slorber Is there a knob or a button I can use to disable `copy-text-to-clipboard`?  Perhaps just `nohoist` in package.json?

**Reply by slorber** [reply]
No there's nothing to disable this.

But this probably hides another problem. You shouldn't need `@babel/plugin-transform-modules-commonjs`, and removing it raises a ton of new errors.

Your site clearly has some weird issues but can't tell what exactly. You should try to remove things one by one until you end up with a very basic Docusaurus site that works. It may be a dependency issue, you should really remove everything until you figure this out, including rollup, custom elements, extra babel plugins...

**Reply by tony** [reply]
@slorber It looks like I'd be better off bootstrapping a fresh docusaurus website from scratch since I'm so many releases behind.

**Reply by slorber** [reply]
yes probably 😅
